File: backend/models/channel.py
from datetime import datetime
from sqlalchemy import String, DateTime, ForeignKey, Enum, Text, Integer, Boolean
from sqlalchemy.orm import Mapped, mapped_column
from database import Base
import enum


class ChannelType(str, enum.Enum):
    text = "text"
    voice = "voice"
    mixed = "mixed"


class Channel(Base):
    __tablename__ = "channels"

    id: Mapped[int] = mapped_column(primary_key=True)
    world_id: Mapped[int] = mapped_column(ForeignKey("worlds.id"))
    name: Mapped[str] = mapped_column(String(100))
    description: Mapped[str] = mapped_column(String(300), default="")
    type: Mapped[ChannelType] = mapped_column(Enum(ChannelType), default=ChannelType.mixed)
    category: Mapped[str] = mapped_column(String(100), default="Général")
    position: Mapped[int] = mapped_column(Integer, default=0)
    is_default: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


class Message(Base):
    __tablename__ = "messages"

    id: Mapped[int] = mapped_column(primary_key=True)
    channel_id: Mapped[int] = mapped_column(ForeignKey("channels.id"))
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"))
    content: Mapped[str] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


class AIMessage(Base):
    __tablename__ = "ai_messages"

    id: Mapped[int] = mapped_column(primary_key=True)
    channel_id: Mapped[int] = mapped_column(ForeignKey("channels.id"))
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"))
    role: Mapped[str] = mapped_column(String(20))  # user | assistant
    content: Mapped[str] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


class CanvasState(Base):
    __tablename__ = "canvas_states"

    id: Mapped[int] = mapped_column(primary_key=True)
    channel_id: Mapped[int] = mapped_column(ForeignKey("channels.id"), unique=True)
    data: Mapped[str] = mapped_column(Text, default="{}")  # Excalidraw JSON
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_by: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=True)

File: mobile/lib/models/channel.dart
class Channel {
  final int id;
  final int worldId;
  final String name;
  final String description;
  final String type;
  final String category;
  final int position;
  final bool isDefault;

  Channel({
    required this.id,
    required this.worldId,
    required this.name,
    required this.description,
    required this.type,
    required this.category,
    required this.position,
    required this.isDefault,
  });

  factory Channel.fromJson(Map<String, dynamic> json) => Channel(
        id: json['id'],
        worldId: json['world_id'],
        name: json['name'],
        description: json['description'] ?? '',
        type: json['type'] ?? 'text',
        category: json['category'] ?? 'Général',
        position: json['position'] ?? 0,
        isDefault: json['is_default'] ?? false,
      );

  String get icon {
    switch (type) {
      case 'voice':
        return '🔊';
      case 'mixed':
        return '🎙️';
      default:
        return '#';
    }
  }
}
